modAionImpl/src/org/aion/zero/impl/sync/TaskImportTrieData.java

```diff
@@ -0,0 +1,141 @@
+package org.aion.zero.impl.sync;
+
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.BlockingQueue;
+import java.util.stream.Collectors;
+import org.aion.base.util.ByteArrayWrapper;
+import org.aion.base.util.Hex;
+import org.aion.mcf.trie.TrieNodeResult;
+import org.aion.zero.impl.AionBlockchainImpl;
+import org.slf4j.Logger;
+
+/**
+ * Processes the received trie nodes that were requested. The thread is shut down once the fast sync
+ * manager indicates that the full trie is the complete.
+ *
+ * @author Alexandra Roatis
+ */
+final class TaskImportTrieData implements Runnable {
+
+    private final Logger log;
+    private final FastSyncManager fastSyncMgr;
+
+    private final AionBlockchainImpl chain;
+    private final BlockingQueue<TrieNodeWrapper> trieNodes;
+
+    /**
+     * Constructor.
+     *
+     * @param log logger for reporting execution information
+     * @param chain the blockchain used by the application
+     * @param trieNodes received trie nodes
+     * @param fastSyncMgr manages the fast sync process and indicates when completeness is reached
+     */
+    TaskImportTrieData(
+            final Logger log,
+            final AionBlockchainImpl chain,
+            final BlockingQueue<TrieNodeWrapper> trieNodes,
+            final FastSyncManager fastSyncMgr) {
+        this.log = log;
+        this.chain = chain;
+        this.trieNodes = trieNodes;
+        this.fastSyncMgr = fastSyncMgr;
+    }
+
+    @Override
+    public void run() {
+        // importing the trie state should be highest priority
+        // since it is usually the longest process (on large blockchains)
+        Thread.currentThread().setPriority(Thread.MAX_PRIORITY);
+
+        while (!fastSyncMgr.isComplete()) {
+            TrieNodeWrapper tnw;
+            try {
+                tnw = trieNodes.take();
+            } catch (InterruptedException ex) {
+                if (!fastSyncMgr.isComplete()) {
+                    log.error("<import-trie-nodes: interrupted without shutdown request>", ex);
```

Comment:
good point!!

Reply:
Similar to `TaskImportBlocks.java` the sync manager will know of an instance of this class and trigger the shutdown when appropriate.

Reply:
I see... is that shutdown just setting fastSyncMgr.isComplete to true?

Reply:
Yep. Since the functionality is a one-time catch-up with the chain use case, it will be permanently disabled after completion. In fact, it also will not be enabled if the initial database is not empty.